system-parachains/asset-hubs/asset-hub-kusama/src/impls.rs

```diff
@@ -0,0 +1,309 @@
+					// Paid amount is fully refunded.
+					Ok(SameOrOther::None) => fungibles::Credit::<T::AccountId, F>::zero(A::get()),
+					// Should never fail as at this point the asset id is always valid and the
+					// refund amount is not greater than paid amount.
+					_ => return Err(InvalidTransaction::Payment.into()),
+				};
+				// No separation for simplicity.
+				// In our case the fees and the tips are deposited to the same pot.
+				// We cannot call [`OnUnbalanced::on_unbalanceds`] since fungibles credit does not
+				// implement `Imbalanced` trait.
+				OU::on_unbalanced(adjusted_paid);
+			}
+			Ok(())
+		}
+	}
+
+	type LiquidityInfoOf<T> =
+		<<T as pallet_transaction_payment::Config>::OnChargeTransaction as OnChargeTransaction<
+			T,
+		>>::LiquidityInfo;
+
+	/// Implements [`OnChargeAssetTransaction`] for [`pallet_asset_conversion_tx_payment`], where
+	/// the asset class used to pay the fee is defined with the `A` type parameter (eg. DOT
+	/// location) and accessed via the type implementing the [`frame_support::traits::fungibles`]
+	/// trait.
+	pub struct SwapCreditAdapter<A, S>(PhantomData<(A, S)>);
+	impl<A, S, T> OnChargeAssetTransaction<T> for SwapCreditAdapter<A, S>
+	where
+		A: Get<S::AssetKind>,
+		S: SwapCredit<
+			T::AccountId,
+			Balance = T::Balance,
+			AssetKind = T::AssetKind,
+			Credit = fungibles::Credit<T::AccountId, T::Assets>,
+		>,
+
+		T: pallet_asset_conversion_tx_payment::Config,
+		T::Fungibles:
+			fungibles::Inspect<T::AccountId, Balance = T::Balance, AssetId = T::AssetKind>,
+		T::OnChargeTransaction:
+			OnChargeTransaction<T, Balance = T::Balance, LiquidityInfo = Option<S::Credit>>,
+	{
+		type AssetId = T::AssetKind;
+		type Balance = T::Balance;
+		type LiquidityInfo = T::Balance;
+
+		fn withdraw_fee(
+			who: &<T>::AccountId,
+			_call: &<T>::RuntimeCall,
+			_dispatch_info: &DispatchInfoOf<<T>::RuntimeCall>,
+			asset_id: Self::AssetId,
+			fee: Self::Balance,
+			_tip: Self::Balance,
+		) -> Result<(LiquidityInfoOf<T>, Self::LiquidityInfo, T::Balance), TransactionValidityError>
+		{
+			let asset_fee = AssetConversion::<T>::quote_price_tokens_for_exact_tokens(
+				asset_id.clone(),
+				A::get(),
+				fee,
+				true,
+			)
+			.ok_or(InvalidTransaction::Payment)?;
+
+			let asset_fee_credit = T::Assets::withdraw(
+				asset_id.clone(),
+				who,
+				asset_fee,
+				Precision::Exact,
+				Preservation::Preserve,
+				Fortitude::Polite,
+			)
+			.map_err(|_| TransactionValidityError::from(InvalidTransaction::Payment))?;
+
+			let (fee_credit, change) = match S::swap_tokens_for_exact_tokens(
+				vec![asset_id, A::get()],
+				asset_fee_credit,
+				fee,
+			) {
+				Ok((fee_credit, change)) => (fee_credit, change),
+				Err((credit_in, _)) => {
+					let _ = T::Assets::resolve(who, credit_in).defensive();
+					return Err(InvalidTransaction::Payment.into())
+				},
+			};
+
+			ensure!(change.peek() == Zero::zero(), InvalidTransaction::Payment);
```

Comment:
DQ but what is the transactional semantic of this function?  
Dont you need to undo everything when it fails? Sounds a bit annoying...

Reply:
This is part of validate and pre_dispatch, so the extrinsic wont be applied if this errors